File: src/sorting/priority_queue.cpp
// 最大優先佇列（CLRS §6.5）：MAXIMUM / EXTRACT-MAX / INCREASE-KEY / INSERT
#include <iostream>
#include <limits>
#include <vector>

void maxHeapify(std::vector<int>& A, int heapSize, int i) {
    int l = 2 * i + 1, r = 2 * i + 2, largest = i;
    if (l < heapSize && A[l] > A[largest]) largest = l;
    if (r < heapSize && A[r] > A[largest]) largest = r;
    if (largest != i) {
        std::swap(A[i], A[largest]);
        maxHeapify(A, heapSize, largest);
    }
}

int heapExtractMax(std::vector<int>& A) {      // O(lg n)：尾補頂、下沉
    int mx = A[0];
    A[0] = A.back();
    A.pop_back();
    maxHeapify(A, A.size(), 0);
    return mx;
}

void heapIncreaseKey(std::vector<int>& A, int i, int key) {   // O(lg n)：上浮
    A[i] = key;
    while (i > 0 && A[(i - 1) / 2] < A[i]) {
        std::swap(A[i], A[(i - 1) / 2]);
        i = (i - 1) / 2;
    }
}

void maxHeapInsert(std::vector<int>& A, int key) {  // 先塞 -∞ 佔位再增鍵
    A.push_back(std::numeric_limits<int>::min());
    heapIncreaseKey(A, A.size() - 1, key);
}

void show(const std::vector<int>& A) {
    for (int v : A) std::cout << v << ' ';
    std::cout << "\n";
}

int main() {
    std::vector<int> A = {16, 14, 10, 8, 7, 9, 3, 2, 4, 1};   // CLRS 圖 6.5
    maxHeapInsert(A, 15);
    std::cout << "INSERT(15) 後: "; show(A);
    std::cout << "EXTRACT-MAX = " << heapExtractMax(A) << "\n之後: "; show(A);
    return 0;
}

File: src/sorting/priority_queue.py
# 最大優先佇列（CLRS §6.5；改寫自官方 max_heap_priority_queue.py，MIT License）
def max_heapify(A, heap_size, i):
    l, r = 2 * i + 1, 2 * i + 2
    largest = i
    if l < heap_size and A[l] > A[largest]:
        largest = l
    if r < heap_size and A[r] > A[largest]:
        largest = r
    if largest != i:
        A[i], A[largest] = A[largest], A[i]
        max_heapify(A, heap_size, largest)


def heap_maximum(A):                     # MAXIMUM：看堆頂，O(1)
    return A[0]


def heap_extract_max(A):                 # EXTRACT-MAX：尾端補頂、下沉修復，O(lg n)
    mx = A[0]
    A[0] = A[-1]
    A.pop()
    max_heapify(A, len(A), 0)
    return mx


def heap_increase_key(A, i, key):        # INCREASE-KEY：增鍵後往上浮，O(lg n)
    assert key >= A[i], "新 key 必須 >= 目前值"
    A[i] = key
    while i > 0 and A[(i - 1) // 2] < A[i]:   # 和父節點比，較大就浮上去
        A[i], A[(i - 1) // 2] = A[(i - 1) // 2], A[i]
        i = (i - 1) // 2


def max_heap_insert(A, key):             # INSERT：先塞 -inf 佔位，再增鍵上浮
    A.append(float("-inf"))
    heap_increase_key(A, len(A) - 1, key)


A = [16, 14, 10, 8, 7, 9, 3, 2, 4, 1]    # CLRS 圖 6.5 的 max-heap
max_heap_insert(A, 15)
print("INSERT(15) 後:", A)
print("EXTRACT-MAX =", heap_extract_max(A))
print("之後:", A)
